### PR TITLE
Replace terminaltables with rich

### PR DESCRIPTION
**Pull request by noprobelm**
## 📝 Description

Relates to #436 

The version of `terminaltables` used by this project has been archived for a few years now and was recently removed from the Arch community repository (which is what led me to embark on this PR). While the project may remain viable for some time to come, I propose the maintainers of `linode-cli` consider switching to [rich](https://github.com/textualize/rich), which offers much greater versatility when rendering rich text to the terminal in Python (even `pip` uses rich for its progress bar functionality).

I have made the necessary changes to the code in this PR.

### Summary of changes:
- Updated `requirements.txt`
  - Added `rich`
  - Removed `terminaltables`
- Updated all instances of `terminaltables.SingleTable` objects throughout the project with equivalent `rich.Table` objects. The following modules were updated under this effort:
  - `arg_helpers.py`
  - `output.py` - Note for this module I also added the `rich.Text` module to enable proper ANSI sequence rendering. Without converting the table cells to `Text` objects, any cell containing ANSI sequences will result in improper rendering of the columns. This should not be applicable to any other module, but could be implemented for uniformity's sake if desired with no impact to functionality.
  - `__init__.py`
  - `plugins/obj/__init__.py`
  - `plugins/obj/helpers.py`
- Updated the `firewall-editor` plugin to replace `terminaltables.PorcelainTable` object with `rich.Table`. The new `rich.Table` is configured to be equivalent to the corresponding `PorcelainTable` it's replacing.
- Updated the unit test `test_output.py` to use `rich.Table`

### Steps
- Run any set of commands associated with generating tables. Among these, make sure to run a command that generates a table with colored text, such as `linode-cli linodes list` to ensure the ANSI escape sequence'd text renders correctly.
- Run `linode-cli` to see the main help message tables
- Run `linode-cli firewall-editor <label>` to see the new firewall editor table.
- Run the `test_output.py` unit test `python -m unittest test_output.py`

### What I tested
- Each of the 4 tests above

### What I **did not** test
- I did  not run the full test suite as I didn't want to destroy my account's configuration nor did I want to make a new account for this purpose. 

### Before/After
- `linode-cli` before
linode-cli-help-before
- `linode-cli` after
linode-cli-help-after
- `linode-cli linodes list` before
linode-cli-linodes-list-before
- `linode-cli linodes list` after
linode-cli-linodes-list-after
- `linode-cli firewall-editor <label>` before
linode-cli-firewall-editor-before
- `linode-cli firewall-editor <label>` after
linode-cli-firewall-editor-after

## Other Notes

I don't understand exactly how the `Makefile` works for this project, but something prevents me from acquiring the necessary baked data (specifically `bake_version`) using `make` in my fork of the project. If I run `make` from this project and copy the `baked_version` file to my fork, I am able to install with `pip`. I did not make any changes to the `Makefile` (or anything else other than what I listed above for that matter), but I thought this would be worth noting in the PR.

I hope you seriously consider merging this PR and/or implementing `rich` on your own. I can't speak highly enough about its performance and options. Thank you for taking the time to consider these changes.

### CODE REVIEW

File: linodecli/plugins/obj/helpers.py [hunk 1]
```diff
@@ -120,13 +120,8 @@ def _denominate(total):
...
-    Returns a terminaltables.SingleTable object with no borders and correct padding
+    Returns a rich.Table object with no borders and correct padding
     """
-    tab = SingleTable(data)
-    tab.inner_heading_row_border = False
-    tab.inner_column_border = False
-    tab.outer_border = False
-    tab.padding_left = 0
-    tab.padding_right = 2
+    tab = Table(show_header=False, show_edge=False, padding=(0, 2, 0, 2))
```
Comment (Dorthu) [inline linodecli/plugins/obj/helpers.py:125]: This differs a bit from the current format, and you need to add the rows in:

```suggestion
    tab = Table(show_header=False, show_edge=False, padding=(0, 2, 0, 2), box=None)
    for row in data:
        row = [Text.from_ansi(str(item)) for item in row]
        tab.add_row(*row)
```

I converted `item` to a string here, but that creates very verbose date strings - some special formatting for them is probably warranted too.

File: linodecli/plugins/obj/__init__.py
Comment (Dorthu) [inline linodecli/plugins/obj/__init__.py:1]: There are several occurrences of `print(tab.table)` left in this file that should be updated

File: linodecli/output.py [hunk 2]
```diff
@@ -134,15 +136,18 @@ def _table_output(
             ),
         )
 
-        tab = SingleTable(content)
+        tab = Table(*content[0])
```
Comment (Dorthu) [inline linodecli/output.py:139]: To preserve existing styling, these arguments are needed:

```suggestion
        tab = Table(*content[0], header_style="", box=box.SQUARE)
```

**Review by Dorthu — COMMENTED** [review]
This is fantastic, and a much-needed update 👍  

I left a few suggestions to preserve the existing style, but ultimately if it changes a bit I don't think it's a huge deal.  I'd be happy to push up changes if you like.

File: linodecli/plugins/obj/helpers.py [hunk 3]
```diff
@@ -4,7 +4,7 @@
 from argparse import ArgumentTypeError
 from datetime import datetime
 
-from terminaltables import SingleTable
+from rich.table import Table
```
Comment (noprobelm) [inline linodecli/plugins/obj/helpers.py:8]: Need to add `rich.Text` import to accommodate @Dorthu suggestion in comment below
```suggestion
from rich.text import Text

```

File: linodecli/plugins/obj/helpers.py (hunk 1 above)
Comment (noprobelm) [reply linodecli/plugins/obj/helpers.py:125]: Can you explain your steps to test this module so I can take a look for myself?

Comment (noprobelm) [reply linodecli/plugins/obj/helpers.py:125]: See my suggestion above to add `rich.Text` import to accommodate this.

**Review by noprobelm — COMMENTED** [review]
Made some changes/comments to adjudicate @Dorthu review

File: linodecli/plugins/obj/helpers.py (hunk 1 above)
Comment (Dorthu) [reply linodecli/plugins/obj/helpers.py:125]: Yeah, my full diff on this file was:

```diff
diff --git a/linodecli/plugins/obj/helpers.py b/linodecli/plugins/obj/helpers.py
index b3a5883..ce247e0 100644
--- a/linodecli/plugins/obj/helpers.py
+++ b/linodecli/plugins/obj/helpers.py
@@ -5,6 +5,7 @@ from argparse import ArgumentTypeError
 from datetime import datetime

 from rich.table import Table
+from rich.text import Text

 from linodecli.plugins.obj.config import DATE_FORMAT

@@ -122,6 +123,9 @@ def _borderless_table(data):
     """
     Returns a rich.Table object with no borders and correct padding
     """
-    tab = Table(show_header=False, show_edge=False, padding=(0, 2, 0, 2))
+    tab = Table(show_header=False, show_edge=False, padding=(0, 2, 0, 2), box=None)
+    for row in data:
+        row = [Text.from_ansi(str(item)) for item in row]
+        tab.add_row(*row)

     return tab
```

I wasn't testing with a script, just installed the cli from source (`pip install -e .` in the root of this repo) and ran some commands; `linode-cli obj ls` was enough to see that no data was displayed, and `linode-cli ls $BUCKET_NAME` to show the contents of a bucket (using the same output formatting).  If you don't have object storage set up, `linode-cli obj mb $BUCKET_NAME` will do that for you and create a new bucket (but I think there's a nominal charge for enabling the service).

**Review by Dorthu — APPROVED** [review]
Tested thoroughly, this looks good 👍

File: tests/unit/test_output.py [hunk 4]
```diff
@@ -280,9 +285,10 @@ def test_truncated_table(self, mock_cli):
         data[0]["cool"] = test_str_truncated
 
         mock_table = io.StringIO()
-        tab = SingleTable([["h1"], [test_str_truncated]])
+        tab = Table("h1")
```
Comment (lgarber-akamai) [inline tests/unit/test_output.py:288]:
```suggestion
        tab = Table("h1", header_style="", box=box.SQUARE)
```

**Review by lgarber-akamai — CHANGES_REQUESTED** [review]
Thanks for the great contribution! The implementation all looks good, although I noticed the altered unit tests were failing on my machine. 

Updating the mock table to use the correct box type seems to have resolved the issue 👍

File: tests/unit/test_output.py [hunk 5]
```diff
@@ -1,7 +1,8 @@
 import contextlib
 import io
 
-from terminaltables import SingleTable
+from rich.table import Table
+from rich import print as rprint
```
Comment (noprobelm) [inline tests/unit/test_output.py:6]:
```suggestion
from rich import box

```

**Review by noprobelm — COMMENTED** [review]
Adding box import to `test_output` unit test

**Review by lgarber-akamai — APPROVED** [review]
Looks great! Thanks again for the excellent contribution 🎉

File: linodecli/plugins/obj/__init__.py [hunk 6]
```diff
@@ -486,9 +487,8 @@ def print_help(parser: ArgumentParser):
         for name, func in sorted(COMMAND_MAP.items())
     ]
 
-    tab = SingleTable(command_help_map)
-    tab.inner_heading_row_border = False
-    print(tab.table)
+    tab = Table(show_header=False)
+    rprint(tab)
```
Comment (noprobelm) [inline linodecli/plugins/obj/__init__.py:491]: This should take care of that failed `pylint` check blocking the merge. I went ahead and

- Made these changes locally
- Checked to make sure the help table populates with `linode-cli obj --help`
- Ran `pylint linodecli` in the root project dir
- For good measure, checked the rest of the code base where the new `rich.Table` objects are to made sure their row content was added  

Should be all good to go now. Please commit this suggestion if you agree with the changes.

```suggestion
    for row in command_help_map:
        tab.add_row(*row)
    rprint(tab)
```

Comment (noprobelm) [reply linodecli/plugins/obj/__init__.py:493]: Committed so you can just re-run the checks if you desire.

**Comment by noprobelm** [comment]
Pushed commits to reorder imports in accordance with `isort` recommendations. `make lint` is now passing for me locally.

**Review by zliang-akamai — APPROVED** [review]

**Review by jriddle-linode — APPROVED** [review]
Really appreciate the hard work @noprobelm !!

**Comment by noprobelm** [comment]
@jriddle-linode Thanks, I'm very happy to contribute! Thank you @Dorthu @lgarber-akamai and @zliang-akamai for walking through this PR with me.

**Pull request merged**